128-bit vector primitive types

TheNumbat opened the pull request:
This includes the front & middle end changes breaking out `vec128` into specific types representing different interpretations of the 128 bits. The more specific types are currently ignored by the backend - a following PR will add typed intrinsics in the backend.

goldfirere commented on ocaml/typing/primitive.ml:
```diff
@@ -293,9 +313,11 @@ let equal_boxed_integer bi1 bi2 =
+  (* For the purposes of layouts/native representations,
+     all 128-bit vector types are equal. *)
+  match bi1, bi2 with
+  | Pvec128 _, Pvec128 _ -> true
```
This implementation does not match what I would expect from the name of the function. I do agree that this behavior is correct for the use cases of the function, so I don't think there's a bug. Maybe `equal_boxed_vector_size`?

TheNumbat replied:
Renamed to `equal_boxed_vector_size`

goldfirere commented on ocaml/typing/predef.ml:
```diff
@@ -88,13 +100,19 @@ and type_option t = newgenty (Tconstr(path_option, [t], ref Mnil))
 and type_string = newgenty (Tconstr(path_string, [], ref Mnil))
 and type_extension_constructor =
-      newgenty (Tconstr(path_extension_constructor, [], ref Mnil))
+newgenty (Tconstr(path_extension_constructor, [], ref Mnil))
```
Accidentally dropped whitespace?

TheNumbat replied:
Fixed

goldfirere commented on ocaml/middle_end/printclambda.ml:
```diff
@@ -61,7 +62,7 @@ let layout (layout : Lambda.layout) =
   | Punboxed_int Pint64 -> ":unboxed_int64"
   | Punboxed_int Pnativeint -> ":unboxed_nativeint"
-  | Punboxed_vector Pvec128 -> ":unboxed_vec128"
+  | Punboxed_vector (Pvec128 ty) -> ":unboxed" ^ (Primitive.vec128_name ty)
```
```suggestion
  | Punboxed_vector (Pvec128 ty) -> ":unboxed_" ^ (Primitive.vec128_name ty)
```

goldfirere reviewed (approved):
Files in `ocaml/` except those in `ocaml/middle_end/`

Ekdohibs commented:
It is unclear to me why we need the vec128 types from lambda onwards. It seems to be that since vec128's are all represented the same way at runtime (they have compatible layouts), then it is the operations on them which need to hold the type information, not the vectors themselves.
Besides, this also raises different questions: what should happen in a vec128 is used with a different type than it was created? It seems this probably should be allowed, because we want to be able to cast between the different representations of vec128 having different number of ints for instance; but this means again that the type information should go on the operations rather than on the values.
Finally, what should the semantics of polymorphic comparison be? Do we want to specify that it is bitwise comparison, even if a vec128 is used to represent a couple of floats? Without this, it is not possible to use the same representation for them as the rest.

TheNumbat commented:
We don't have to track the types from lambda onward, but @mshinwell suggested doing so to make middle-end optimizations (e.g. constant folding) possible in the future. Currently, they still won't be possible since the (intrinsic) operations won't get translated until after the middle end, but we'd like to change that sometime soon. 

I'm planning on supporting casts using a new operation (https://github.com/ocaml-flambda/flambda-backend/blob/simd-intrins/backend/cmm.mli#L187) that works like the float_of_bits/bits_of_floats path.
Overall, though, I agree it makes more sense for the types to be enforced by the operations rather than the values - they aren't preserved in Cmm, so at that the operation types are only enforced by the `external ... [@@builtin]` declarations in library code.

As for polymorphic comparison, all vectors currently use `abstract_tag`, so comparing them will raise. I think it's fine if we want to specify that it's always a bitwise comparison, but we'd have to add a new tag. Should I do that in another PR?

Ekdohibs commented:
If we have a specific cast operation, it should be good. Concerning comparison, raising is consistent and well-specified, so I'm good with this.

Ekdohibs reviewed (changes requested):
We must decide whether two different vec128 types are supposed to be compatible or not. However, the current implementation sometimes makes them incompatible and sometimes compatible, which could cause compilation errors if I'm not mistaken.
For instance, flambda2_types supposes these types are incompatible, while the code for comparing constants assumes they are equal. If we want the types to be fully incompatible, we would need to change these places, including in Lambda (though we can drop the distinction in cmm); if we want them to be compatible, we might want to still track the type a bit (for instance with the coercion machinery, although I'm unsure if it is robust enough for that; maybe @lukemaurer could weigh in on that?).
I'm open to both options, but we need to be consistent if we want to avoid the risk of compilation failures, for instance when using GADTs and vector types.

TheNumbat commented on backend/cmm_helpers.ml:
```diff
@@ -718,15 +718,16 @@ let rec unbox_float dbg =
       match Cmmgen_state.structured_constant_of_sym s.sym_name with
-      | Some (Uconst_vec128 { low; high }) ->
+      | Some (Uconst_vec128 { ty = cty; low; high }) ->
+        assert (Lambda.equal_vec128 ty cty);
```
Now ignored

TheNumbat replied:
I put the check back; is it ok now?

TheNumbat commented:
I've addressed the comments, but not sure how we should deal with compatibility. I think we at least want the constants to be compatible, since (e.g.) zero is exactly the same for all of the types. 

It looks like the only places that treat the types as distinct are `meet_and_join.ml` and `provers.ml` - should I make these cases compatible as well? It's less clear whether they should be, since they'd have to choose one of the types to return. 
(However, at that point there would be no current need to track the types in the middle end at all; just that doing so should be useful for future optimizations.)

TheNumbat commented:
Ok, I tried making the vector types a lattice - I don't seem to have broken anything, but I'm very not-confident that it's correct.

Ekdohibs commented on backend/cmm_helpers.ml:
```diff
@@ -718,15 +718,17 @@ let rec unbox_float dbg =
       match Cmmgen_state.structured_constant_of_sym s.sym_name with
-      | Some (Uconst_vec128 { ty = _; low; high }) ->
+      | Some (Uconst_vec128 { ty; low; high }) ->
+        (* CR mslater for ekdohibs: is this now always true? *)
```
Unfortunately I think not, due to the presence of GADTs.

Ekdohibs commented on middle_end/flambda2/numbers/vector_types.ml:
```diff
@@ -60,56 +60,110 @@ end
+    | Unknown128, _ -> Unknown128
+    | _, Unknown128 -> Unknown128
+    | Any128, _ -> r
+    | _, Any128 -> l
```
Maybe `| Any128, vty | vty, Any128 -> vty` would be better?

Ekdohibs commented on middle_end/flambda2/numbers/vector_types.ml:
```diff
@@ -60,56 +60,110 @@ end
+    | Float32x4, Float32x4 -> Float32x4
+    | Float32x4, _ -> Unknown128
+    | Float64x2, Float64x2 -> Float64x2
+    | Float64x2, _ -> Unknown128
```
Style-wise, the rest of the code usually lists all identical cases, and then a single `(Int8x16 | Inx16x8 | ...), _` at the end to make sure the pattern matching is not fragile.

Ekdohibs commented:
I just realized that different types of identically-sized vectors should not be different kinds (in flambda2), but different subkinds of the same kind if we want to allow sharing constants.
There might be a couple hidden complexities with this, as I think for now only values have useful subkinds, and in particular naked numbers do not. I'll try to think a bit about the best way to proceed.

TheNumbat commented:
Hmm, I see what you mean about naked number subkinds. Are the `Naked_vector` kinds currently treated as distinct? I think for  sharing constants, they all already map to a `head_of_kind_naked_vec128` and can meet/join for the same sizes - but unsure if they're treated differently elsewhere. Is using a sub-kind something I should do in this PR?

Ekdohibs commented:
Well, they're not treated as distinct per se by meet/join, but there is code that expects that if you take only the kind of the variable (and not its subkind), it will be the same everywhere, so two variables with compatible layouts should have compatible flambda2 kinds. However, this might cause problems with immediate/immediate64 being a sublayout of int64 (for now we handler anything immediate like we would handle a value); so there might be deeper changes to do in flambda2 for that; I'll try to ask Vincent about this.

lthls commented:
I looked a bit at the issue of Flambda 2 types.
Basically what we want in the types is a single `Naked_vec128` kind (plus new ones for the other sizes). The subkinds are not relevant at the moment: the only information we track is in the form of finite sets of constants, so we don't really care about mixing `Int8x16` and `Float64x2` (and we want to mix them for sharing constants).
The primitives will of course depend on the actual layout, as adding vectors of 16 integers or vectors of 2 floats will not produce the same result, but in the type and kind systems we don't need the extra precision.

Where subkinds could be useful is if we want to track relations between parts of a vector. In pseudo-code:
```ocaml
let f (x : #int64) (y : #int64) =
  let v : #int64x2 = { high = x; low = y } in
  let z = v.high in
  assert (z == x);
  ...
```
If we want to be able to prove the assertion (and remove the branch), we need to track vectors with a type that looks more like the type of records than that of naked numbers. It's doable, but would be a bit of work to implement correctly so I'd suggest to wait until we have an actual need for this kind of precision.

TheNumbat commented:
Alright, I've reverted the specific type tracking in flambda2 kinds/types - the flambda2 changes are now mostly just refactoring. 
Currently the vector types are opaque, so I agree the component tracking can wait.

TheNumbat commented:
Sounds good, shall we merge this now?
